## Python Built-in Functions

## The `help()` Function

In [117]:
help(len)

Help on built-in function len in module builtins:

len(obj, /)
    Return the number of items in a container.



In [118]:
help(round)

Help on built-in function round in module builtins:

round(number, ndigits=None)
    Round a number to a given precision in decimal digits.
    
    The return value is an integer if ndigits is omitted or None.  Otherwise
    the return value has the same type as the number.  ndigits may be negative.



In [119]:
help(map)

Help on class map in module builtins:

class map(object)
 |  map(func, *iterables) --> map object
 |  
 |  Make an iterator that computes the function using arguments from
 |  each of the iterables.  Stops when the shortest iterable is exhausted.
 |  
 |  Methods defined here:
 |  
 |  __getattribute__(self, name, /)
 |      Return getattr(self, name).
 |  
 |  __iter__(self, /)
 |      Implement iter(self).
 |  
 |  __next__(self, /)
 |      Implement next(self).
 |  
 |  __reduce__(...)
 |      Return state information for pickling.
 |  
 |  ----------------------------------------------------------------------
 |  Static methods defined here:
 |  
 |  __new__(*args, **kwargs) from builtins.type
 |      Create and return a new object.  See help(type) for accurate signature.



In [120]:
help(zip)

Help on class zip in module builtins:

class zip(object)
 |  zip(*iterables) --> A zip object yielding tuples until an input is exhausted.
 |  
 |     >>> list(zip('abcdefg', range(3), range(4)))
 |     [('a', 0, 0), ('b', 1, 1), ('c', 2, 2)]
 |  
 |  The zip object yields n-length tuples, where n is the number of iterables
 |  passed as positional arguments to zip().  The i-th element in every tuple
 |  comes from the i-th iterable argument to zip().  This continues until the
 |  shortest argument is exhausted.
 |  
 |  Methods defined here:
 |  
 |  __getattribute__(self, name, /)
 |      Return getattr(self, name).
 |  
 |  __iter__(self, /)
 |      Implement iter(self).
 |  
 |  __next__(self, /)
 |      Implement next(self).
 |  
 |  __reduce__(...)
 |      Return state information for pickling.
 |  
 |  ----------------------------------------------------------------------
 |  Static methods defined here:
 |  
 |  __new__(*args, **kwargs) from builtins.type
 |      Create and ret

In [121]:
help(enumerate)

Help on class enumerate in module builtins:

class enumerate(object)
 |  enumerate(iterable, start=0)
 |  
 |  Return an enumerate object.
 |  
 |    iterable
 |      an object supporting iteration
 |  
 |  The enumerate object yields pairs containing a count (from start, which
 |  defaults to zero) and a value yielded by the iterable argument.
 |  
 |  enumerate is useful for obtaining an indexed list:
 |      (0, seq[0]), (1, seq[1]), (2, seq[2]), ...
 |  
 |  Methods defined here:
 |  
 |  __getattribute__(self, name, /)
 |      Return getattr(self, name).
 |  
 |  __iter__(self, /)
 |      Implement iter(self).
 |  
 |  __next__(self, /)
 |      Implement next(self).
 |  
 |  __reduce__(...)
 |      Return state information for pickling.
 |  
 |  ----------------------------------------------------------------------
 |  Static methods defined here:
 |  
 |  __new__(*args, **kwargs) from builtins.type
 |      Create and return a new object.  See help(type) for accurate signature.



In [122]:
help(filter)

Help on class filter in module builtins:

class filter(object)
 |  filter(function or None, iterable) --> filter object
 |  
 |  Return an iterator yielding those items of iterable for which function(item)
 |  is true. If function is None, return the items that are true.
 |  
 |  Methods defined here:
 |  
 |  __getattribute__(self, name, /)
 |      Return getattr(self, name).
 |  
 |  __iter__(self, /)
 |      Implement iter(self).
 |  
 |  __next__(self, /)
 |      Implement next(self).
 |  
 |  __reduce__(...)
 |      Return state information for pickling.
 |  
 |  ----------------------------------------------------------------------
 |  Static methods defined here:
 |  
 |  __new__(*args, **kwargs) from builtins.type
 |      Create and return a new object.  See help(type) for accurate signature.



## The `map()` function

In [33]:
numbers  = [4, 8, 9, 13, 15]

cubes  = [number**3 for number in numbers]

print(cubes)

[64, 512, 729, 2197, 3375]


In [9]:
def cube(number):
    return number**3

In [10]:
print(map(cube,numbers))

In [11]:
print(list(map(cube,numbers)))

[64, 512, 729, 2197, 3375]


In [12]:
animals  = ['cat','dog','zebra','donkey','cobra','eagle','elephant','lion']

print(map(len,animals))

In [13]:
print(list(map(len,animals)))

[3, 3, 5, 6, 5, 5, 8, 4]


In [14]:
def capitalize(string):
    return string.upper()

In [15]:
print(list(map(capitalize,animals)))

['CAT', 'DOG', 'ZEBRA', 'DONKEY', 'COBRA', 'EAGLE', 'ELEPHANT', 'LION']


In [21]:
def cap(string):
    return string.title()

In [22]:
print(list(map(cap,animals)))

['Cat', 'Dog', 'Zebra', 'Donkey', 'Cobra', 'Eagle', 'Elephant', 'Lion']


## The `filter()` function
Filter function runs through the iterable automatically, so there would be a need to invoke a loop

In [23]:
animals  = ['cat','dog','zebra','donkey','cobra','eagle','elephant','lion','monkey','giraffe']

long_words = [animal for animal in animals if len(animal) >= 5]

print(long_words)

['zebra', 'donkey', 'cobra', 'eagle', 'elephant', 'monkey', 'giraffe']


In [24]:
def is_long_animal(animal):
    return len(animal) >= 5

In [25]:
print(filter(is_long_animal, animals))

In [26]:
print(list(filter(is_long_animal, animals)))

['zebra', 'donkey', 'cobra', 'eagle', 'elephant', 'monkey', 'giraffe']


## The `Lambda` function

A Lambda function is simply an anonymous function (a function without a name)

In [35]:
metals = ['gold', 'iron', 'silver', 'platinum', 'palladium']

In [39]:
print(list(filter(lambda metal: len(metal) > 5, metals)))

['silver', 'platinum', 'palladium']


In [40]:
print(list(filter(lambda word: 'p' in word, metals)))

['platinum', 'palladium']


In [43]:
print(list(map(lambda word: word.count('l'),metals)))

[1, 0, 1, 1, 2]


In [45]:
print(list(filter(lambda word: word.count('l'),metals)))

['gold', 'silver', 'platinum', 'palladium']


In [46]:
print(list(map(lambda val: val.replace('s','$'), metals)))

['gold', 'iron', '$ilver', 'platinum', 'palladium']


## The `all()` and `any()` function 

In [47]:
print(all([True]))

True


In [49]:
print(all([True,True]))

True


In [50]:
print(all([True,True,False]))

False


In [51]:
#For Numbers, 0 equates to a False Bolean, while other number equate to True
print(all([1,2,3]))

True


In [58]:
print(all([1,2,3,0]))

False


In [53]:
print(all(['a','b']))

True


In [54]:
print(all(['a','b','']))

False


In [56]:
#an empty list evaluates to true
print(all([]))

True


In [60]:
print(any([True,False]))

True


In [61]:
print(any([False,False]))

False


In [62]:
print(any([0,1]))

True


In [63]:
print(any([0]))

False


In [64]:
print(any([" ",""]))

True


In [65]:
print(any(['']))

False


In [66]:
print(any([]))

False


## The `min()` and `max()` functions

In [67]:
print(max([3,4,5,7]))

7


In [68]:
print(max([3,9,23,5]))

23


In [69]:
print(min([3,9,23,5]))

3


In [70]:
print(max(['D','Z','K']))

Z


In [71]:
print(min(['D','Z','K']))

D


## The `sum()` function

In [72]:
print(sum([2,9]))

11


In [73]:
print(sum([2,9.8,-13.78,1000]))

998.02


## The `dir()` function

In [76]:
print(dir([]))

['__add__', '__class__', '__contains__', '__delattr__', '__delitem__', '__dir__', '__doc__', '__eq__', '__format__', '__ge__', '__getattribute__', '__getitem__', '__gt__', '__hash__', '__iadd__', '__imul__', '__init__', '__init_subclass__', '__iter__', '__le__', '__len__', '__lt__', '__mul__', '__ne__', '__new__', '__reduce__', '__reduce_ex__', '__repr__', '__reversed__', '__rmul__', '__setattr__', '__setitem__', '__sizeof__', '__str__', '__subclasshook__', 'append', 'clear', 'copy', 'count', 'extend', 'index', 'insert', 'pop', 'remove', 'reverse', 'sort']


In [79]:
help(dir)

Help on built-in function dir in module builtins:

dir(...)
    dir([object]) -> list of strings
    
    If called without an argument, return the names in the current scope.
    Else, return an alphabetized list of names comprising (some of) the attributes
    of the given object, and of attributes reachable from it.
    If the object supplies a method named __dir__, it will be used; otherwise
    the default dir() logic is used and returns:
      for a module object: the module's attributes.
      for a class object:  its attributes, and recursively the attributes
        of its bases.
      for any other object: its attributes, its class's attributes, and
        recursively the attributes of its class's base classes.



In [80]:
print(dir('bologna'))

['__add__', '__class__', '__contains__', '__delattr__', '__dir__', '__doc__', '__eq__', '__format__', '__ge__', '__getattribute__', '__getitem__', '__getnewargs__', '__gt__', '__hash__', '__init__', '__init_subclass__', '__iter__', '__le__', '__len__', '__lt__', '__mod__', '__mul__', '__ne__', '__new__', '__reduce__', '__reduce_ex__', '__repr__', '__rmod__', '__rmul__', '__setattr__', '__sizeof__', '__str__', '__subclasshook__', 'capitalize', 'casefold', 'center', 'count', 'encode', 'endswith', 'expandtabs', 'find', 'format', 'format_map', 'index', 'isalnum', 'isalpha', 'isascii', 'isdecimal', 'isdigit', 'isidentifier', 'islower', 'isnumeric', 'isprintable', 'isspace', 'istitle', 'isupper', 'join', 'ljust', 'lower', 'lstrip', 'maketrans', 'partition', 'replace', 'rfind', 'rindex', 'rjust', 'rpartition', 'rsplit', 'rstrip', 'split', 'splitlines', 'startswith', 'strip', 'swapcase', 'title', 'translate', 'upper', 'zfill']


The methods that begin and end with the underscores are called Dunda methods.

In [81]:
print(len('bologna'))

7


In [83]:
print('bologna'.__len__())

7


In [84]:
'lo' in 'bologna'

True

In [87]:
print('bologna'.__contains__ ('lo'))

True


In [88]:
print('pasta' + 'meat')

pastameat


In [92]:
print('pasta'.__add__("meat"))

pastameat


In [96]:
print(dir({}))

['__class__', '__contains__', '__delattr__', '__delitem__', '__dir__', '__doc__', '__eq__', '__format__', '__ge__', '__getattribute__', '__getitem__', '__gt__', '__hash__', '__init__', '__init_subclass__', '__iter__', '__le__', '__len__', '__lt__', '__ne__', '__new__', '__reduce__', '__reduce_ex__', '__repr__', '__reversed__', '__setattr__', '__setitem__', '__sizeof__', '__str__', '__subclasshook__', 'clear', 'copy', 'fromkeys', 'get', 'items', 'keys', 'pop', 'popitem', 'setdefault', 'update', 'values']


## The `format()` Function

<p> The format function is used for string interpolation: a way to embed variables into strings. Thus, the print statements are much more powerful with string interpolation.</p>


In [2]:
number = 0.123456789

In [3]:
print(format(number, 'f'))

0.123457


In [4]:
print(type(format(number, 'f')))

<class 'str'>


In [5]:
print(format(number, '.2f'))

0.12


In [6]:
print(format(number, '.1f'))

0.1


In [7]:
print(format(number, '.3f'))

0.123


In [8]:
print(format(0.5, '%'))

50.000000%


In [12]:
print(format(0.5, '.0%'))

50%


In [108]:
print(format(0.5,'f'))

0.500000


In [110]:
print(format(123567, ','))

123,567


In [111]:
print(format(87656789, ','))

87,656,789


In [24]:
def product(a,b):
    return a*b

In [25]:
numbers = [3,5]
number = [3,5]

In [26]:
print(product(numbers))

TypeError: product() missing 1 required positional argument: 'b'

In [27]:
#*arg Unpacking operators
print(product(*numbers))

15


## Python `f-strings`

<p> The python F-string is similar to the `format` function, it is used for efficient string interpolation: a way to embed variables into strings. Thus, the print statements are much more powerful with string interpolation.</p>

<p> Formatted strings literals, also known as f-strings, are a highly practical method for string interpolation. They use curly braces `{}` as variable placeholders. We would go over 4 applications of f-strings in python more efficiently.</p>

In [29]:
age = 24

print(f'John is {age} years old today.')

John is 24 years old today.


### 1. Format Large numbers
When working with large numbers, it is better to use thousands seperators for readability. F-strings make it quite simple to place these seperators appropriately.

In [33]:
number = 345780679087

print(f'The value of the number is {number}')

The value of the number is 345780679087


Lets place the seperators and see the difference.

In [34]:
print(f'The value of the number is {number: ,d}')

The value of the number is  345,780,679,087


### 2. Format dates
There are various ways to represent dates in our scripts. It depends on the geographical location or just your preferences.
We can place dates in f-strings without formatting just like any other variable.

In [35]:
from datetime import datetime

today = datetime.today().date()

print(f'Today is {today}')

Today is 2021-07-19


In [36]:
print(f'Today is {today:%B %d, %Y}')

Today is July 19, 2021


if you live in a country where the month is written before the day, you can use the following:

In [41]:
print(f'Today is {today:%m-%d-%Y}')

Today is 07-19-2021


### Padding Numbers

In some cases, numbers are written with leading zeros in order to have the same number of digits for all numbers. Typical use cases might be product numbers or id Numbers.

We can place any leading zeros for a variable in f-strings.

In [43]:
a = 4

b = 123

print(f"Product numbers are: \n{a:03} \n{b:03}")

Product numbers are: 
004 
123


`"b:03"` indicates that there will be a total of 3 digits and leading spaces will be padded with zero. In case of a one-digit number, we have 2 leading zeros. if it was `b:04`, the number would be written as 0004 and 0123.

### 4. Write Expressions
The f-strings also allow for using expressions in variable placeholders. These expressions can involve function executions. It is a handy feature because we do not need to create variables for values that are used only once.



In [45]:
from datetime import datetime, timedelta

today = datetime.today().date()

print(f"The test was 5 days ago which is {today - timedelta(days=5)}")

The test was 5 days ago which is 2021-07-14


In [47]:
lisst = [23,-1,89,100, 67]

print(f"The list contains {len(lisst)} items.")

The list contains 5 items.
